File: 05/haskell/main.hs
-- The core algorithm here involves generating all points and then running some manipulation over them.
-- This takes a ton of memory and won't scale. We might want to think of another algorithm that scales.
import Data.Either
import Data.List
import System.IO
import Text.ParserCombinators.Parsec

-- Data types
type Point = (Int, Int)

data Line = Line Point Point deriving (Show, Eq)

-- BEGIN PARSING CODE --

lineParser :: CharParser () Main.Line
lineParser = toLine <$> sepBy pointParser (string " -> ")

pointParser :: CharParser () Point
pointParser = toPoint <$> sepBy number (char ',')

number :: CharParser () Int
number = read <$> many1 digit

toPoint :: [Int] -> Point
toPoint [x, y] = (x, y)
toPoint _ = error "Invalid point"

toLine :: [Point] -> Main.Line
toLine [p1, p2] = Line p1 p2
toLine _ = error "Invalid line"

processInput :: String -> [Main.Line]
processInput input = rights $ map (parse lineParser "Invalid Command") $ lines input

-- END PARSING CODE --

sign :: Int -> Int
sign x
  | x < 0 = -1
  | x > 0 = 1
  | otherwise = 0

generatePoints :: Main.Line -> [Point]
generatePoints (Main.Line (x1, y1) (x2, y2)) = [(x1 + d * xDir, y1 + d * yDir) | d <- [0 .. lineLength]]
  where
    dX = x2 - x1
    dY = y2 - y1
    xDir = sign dX
    yDir = sign dY
    lineLength = max (abs dX) (abs dY)

nonDiagonal :: Main.Line -> Bool
nonDiagonal (Main.Line (x1, y1) (x2, y2)) = x1 == x2 || y1 == y2

solution i = (length . filter ((> 1) . length) . group . sort) $ filter nonDiagonal (processInput i) >>= generatePoints

solution2 i = (length . filter ((> 1) . length) . group . sort) $ processInput i >>= generatePoints

main = do
  handle <- openFile "input.txt" ReadMode
  contents <- hGetContents handle
  print (solution2 contents)
  hClose handle

File: 05/python/main.py
def get_lines(text_lines):
    lines = []
    for txt in text_lines:
        [start, end] = txt.split(' -> ')
        [x1, y1] = start.split(',')
        [x2, y2] = end.split(',')
        lines.append(((int(x1), int(y1)), (int(x2), int(y2))))
    return lines

def is_non_diagonal(line):
    return line[0][0] == line[1][0] or line[0][1] == line[1][1]


def generate_points(line):
    if line[0][0] == line[1][0]:
        return [(line[0][0], y) for y in range(line[0][1], line[1][1] + 1)]
    elif line[0][1] == line[1][1]:
        return [(x, line[0][1]) for x in range(line[0][0], line[1][0] + 1)]
    else:
        y_dir = 1 if line[0][1] < line[1][1] else -1 
        points = [(line[0][0] + d, line[0][1] + d * y_dir) for d in range(0, line[1][0] - line[0][0] + 1)]
        return points

def intersection_points(l1, l2):
    p1 = set(generate_points(l1))
    p2 = set(generate_points(l2))
    return p1.intersection(p2)

def solution(text_lines):
    lines = [sorted(x) for x in get_lines(text_lines)]
    points = [generate_points(x) for x in lines]
    # print(points)
    point_set = set()
    intersect_set = set()
    for ps in points:
        for point in ps:
            if point in point_set:
                intersect_set.add(point)
            point_set.add(point)
    
    return len(intersect_set)

with open('input.txt') as f:
    lines = f.readlines()
    print(solution(lines))